library/std/src/sys_common/lazy_box.rs

```diff
@@ -0,0 +1,83 @@
+#![allow(dead_code)] // Only used on some platforms.
+
+// This is used to wrap pthread {Mutex, Condvar, RwLock} in.
+
+use crate::marker::PhantomData;
+use crate::ops::{Deref, DerefMut};
+use crate::ptr::null_mut;
+use crate::sync::atomic::{
+    AtomicPtr,
+    Ordering::{AcqRel, Acquire},
+};
+
+pub(crate) struct LazyBox<T: LazyInit> {
+    ptr: AtomicPtr<T>,
+    _phantom: PhantomData<T>,
+}
+
+pub(crate) unsafe trait LazyInit {
+    /// This is called before the box is allocated, to provide the value to
+    /// move into the new box.
+    ///
+    /// It might be called more than once per LazyBox, as multiple threads
+    /// might race to initialize it concurrently, each constructing and initializing
+    /// their own box. (All but one of them will be destroyed right after.)
+    fn new() -> Self;
+    /// This is called after allocating the box, when the address of `self`
+    /// won't change anymore.
+    unsafe fn init(&mut self);
+    /// This is called before deallocating the box, with the same address for
+    /// `self` as used in `init()`.
+    unsafe fn destroy(&mut self);
+}
+
+impl<T: LazyInit> LazyBox<T> {
+    pub const fn new() -> Self {
+        Self { ptr: AtomicPtr::new(null_mut()), _phantom: PhantomData }
+    }
+
+    unsafe fn get_pointer(&self) -> *mut T {
```

Comment:
Should this be `#[inline]`? We need to double-check where the inlining boundary is with regards to other crates using `std`.